### DISCUSSION TITLE
Set

### DISCUSSION THREAD

**Discussion by vumrra** [subak-ba/cs-study#35] [자료 구조]
# Set

- Set은 순서가 없고 중복을 허용하지 않는 자료구조이다.
- 집합과 같은 개념으로 생각할 수 있다.
- 자바에서는 Set 이라는 인터페이스의 3가지 구현체가 있다.

## Set의 구현체

HashSet

- Hash 알고리즘을 이용한다.

TreeSet

- 이진 탐색 트리를 사용하여 정렬 기능을 재공한다.

LinkedHashSet

- Set에 순서를 부여해준다.

## Set을 쓰는 이유

- Set의 가장 큰 특징은 순서가 없고(비순차적) 중복을 허용하지 않는다는 것이다.
- 위 특징들을 List 같은 다른 자료구조로 구현할 수 있지만 명백한 속도차이가 난다.

- 집합 관련 문제를 해결할 때
- 중복 처리를 해야할때
- 빠르게 찾아야할 때

- 위의 상황에서 Set을 사용한다

## Hash?

- HashSet을 설명하기 전에 hash알고리즘에 대해 간단하게 설명하겠습니다.
- hash는 값을 저장, 조회하는 알고리즘 중 가장 빠른 알고리즘이다.
- key, value 형태로 데이터를 저장한다.
- 동작 방식은 저장하려는 값을 어떤 값과도 중복되지 않는 해시코드로 변환하여 해당 코드의 메모리 위치에 값을 저장하는 방식이다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/7681b4f0-1e01-4818-aa0a-0736473a8ba9)

- 저장하려는 key를 일련의 공식(hash function)을 통해 해시코드로 변한된 값이 저장되는 위치이다.
- 그리하여 조회시에 key만 알고 있으면 해시코드로 변환하여 코드로 한번에 주소에 접근하여 값을 가져올 수 있다.

- 즉 저장과 조회의 시간복잡도는 O(1) 이라는 수치가 나온다.

## HashSet

- HashSet은 hash 알고리즘 기반으로 동작한다
- set은 비선형 자료구조이기 때문에 비순차적으로 값이 저장되고 조회된다.
- 또한 값의 중복을 허용하지 않는다, hash 알고리즘을 사용하기 때문에 저장하려는 값 자체가 메모리 주소가 된다.
    - 그렇기 때문에 값이 똑같은 데이터를 중복 저장하려 할 수 없고 저장하려하면 기존의 값을 대체한다.

```java

public class Main {
    public static void main(String[] args) throws Exception {

        Set set = new HashSet();
        set.add("1최장우");
        set.add("4김원욱");
        set.add("3배선후");
        set.add("2신희성");

        for (Object o : set) {
            System.out.println("o = " + o);
        }
    }
}

/*
	o = 3배선후
	o = 4김원욱
	o = 1최장우
	o = 2신희성
*/

// 비순차적으로 저장, 조회
```

## TreeSet

- TreeSet은 레드블랙트리를 기반으로 동작한다
- 레드블랙트리는 이진트리의 개념에 더하여 한 노드를 기준으로 더 작은 값은 왼쪽, 큰 값은 오른쪽에 저장하는 개념을 추가한 트리이다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/dc44d800-b24a-4503-839c-ef431173810a)

- hashset, treeset은 저장, 조회의 방법만 다른 뿐 기본 개념을 동일하다
- 하지만 구현의 차이로 인해 시간복잡도에 대해서는 차이가 있다.

- 레드블랙트리의 시간복잡도인 저장과 조회에 O(log n)이 걸린다.
- 해당 시간복잡도도 다른 자료구조와 비교해봤을 때는 빠른 편이지만 hashset 보다는 느리다.
- 그렇기 때문에 Set을 사용할 때는 구현체는 대부분 hashset을 많이 사용한다.

```java
public class Main {
    public static void main(String[] args) throws Exception {

        Set set = new TreeSet();
        set.add("1최장우");
        set.add("4김원욱");
        set.add("3배선후");
        set.add("2신희성");

        for (Object o : set) {
            System.out.println("o = " + o);
        }
    }
}

/*
	o = 1최장우
	o = 2신희성
	o = 3배선후
	o = 4김원욱
*/

// 비순차적이지만 레드블랙트리 기반으로 동작하기 때문에 정렬되어 저장된 값을 출력하는 모습니다.
```

## LinkedHashSet

- LinkedHashSet은 hash 알고리즘을 사용하여 저장, 조회를 하되 값을 이중연결리스트에 저장하여 순서를 보장해주는 구현체이다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/efd4365e-66a7-4ff7-a10d-453629a6e73e)

- LinkedHashSet의 내부를 보면 HashSet을 상속받아 hash 알고리즘을 사용한다는 것을 알 수 있다.
- LinkedHashSet도 hash알고리즘을 사용하기 때문에 시간복잡도는 O(1)로 빠르다.

```java
public class Main {
    public static void main(String[] args) throws Exception {

        Set set = new LinkedHashSet();
        set.add("1최장우");
        set.add("4김원욱");
        set.add("3배선후");
        set.add("2신희성");

        for (Object o : set) {
            System.out.println("o = " + o);
        }
    }
}

/*
	o = 1최장우
	o = 4김원욱
	o = 3배선후
	o = 2신희성
*/

// 저장한 순서대로 순차적으로 저장, 조회되는 모습니다.
```

## 실제 사용 예시

- 개인 프로젝트인 GSM GOGO에서 팀을 등록하는 비즈니스 로직 중 일부입니다.
- 한 팀에 중복된 유저가 등록될 수 없으므로 request로 온 유저들의 id를 순회하며 HashSet 자료구조에 저장합니다.
- Set 자료구조에 값을 저장하면 중복되지 않은 값이면 1, 중복되는 값이면 0을 반환합니다.
- 해당 특성을 활용하여 유저들의 중복 저장을 막았습니다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/b5efa477-079e-4ec2-b385-2e1ae81b7d5a)

**Comment by onee-only** [comment]
제가 이해한 바로는 Set는 자료에 중복이 없어야 하는 것으로 보이는데, 
TreeSet에 있는 자료 사진에 6이라는 값이 중복 되어 있는 것 같아요.

**Comment by onee-only** [comment]
> - 레드블랙트리는 이진트리의 개념에 더하여 한 노드를 기준으로 더 작은 값은 왼쪽, 큰 값은 오른쪽에 저장하는 개념을 추가한 트리이다.

설명하고 있는 개념은 이진 탐색 트리에 더 가까운 것 같아요.
레드 블랙 트리는 이진 탐색 트리가 치우치지 않도록 스스로 균형을 맞추는 [자가 균형 이진 탐색 트리](https://ko.wikipedia.org/wiki/%EC%9E%90%EA%B0%80_%EA%B7%A0%ED%98%95_%EC%9D%B4%EC%A7%84_%ED%83%90%EC%83%89_%ED%8A%B8%EB%A6%AC)라고 정의할 수 있어요.
